File: solutions/911/main.cpp
/* ************************************************************************
> File Name:     main.cpp
> Author:        Sean
> Project:    LeetCode
> Created Time:  Fri Dec 10 20:42:05 2021
> Description:   
 ************************************************************************/
#include <vector> 
#include <unordered_map> 
#include <gtest/gtest.h> 

using namespace std; 

class TopVotedCandidate {
private:
    vector<int> times;
    unordered_map<int, int> totalVoteMap;
    unordered_map<int, int> latestVoteMap;
    vector<int> result;
public:
    TopVotedCandidate(vector<int>& persons, vector<int>& times) :
    times(times)
    {
        for (int i = 0; i < times.size(); ++i) {
            if (totalVoteMap.find(persons[i]) == totalVoteMap.end()) totalVoteMap[persons[i]] = 0;
            if (latestVoteMap.find(persons[i]) == latestVoteMap.end()) latestVoteMap[persons[i]] = 0;
            totalVoteMap[persons[i]]++;
            latestVoteMap[persons[i]] = times[i];

            int mvp = 0, mv = INT_MIN;
            for (auto& pv : totalVoteMap) {
                if (pv.second > mv) {
                    mv = pv.second;
                    mvp = pv.first;
                }
                else if (pv.second == mv) {
                    if (latestVoteMap[pv.first] > latestVoteMap[mvp]) mvp = pv.first;
                }
            }
            result.push_back(mvp);
        }
    }

    int q(int t) {
        int left = 0, right = times.size()-1;
        int mid;
        while (left <= right) {
            mid = left + (right - left) / 2;
            if (times[mid] < t) left = mid + 1;
            else if (times[mid] > t) right = mid - 1;
            else return result[mid];
        }
        if (times[mid] > t) mid--;
        return result[mid];
    }
};

TEST(Test911, SimpleTest) {
    vector<int> persons{
        0, 1, 1, 0, 0, 1, 0
    };
    vector<int> times{
        0, 5, 10, 15, 20, 25, 30
    };
    vector<int> queries{
        3, 12, 25, 15, 24, 8
    };
    vector<int> ans{
        0, 1, 1, 0, 0, 1
    };
    TopVotedCandidate s(persons, times); 
    for (int i = 0; i < queries.size(); ++i) {
        EXPECT_EQ(s.q(queries[i]), ans[i]);
    }
}

int main() {
    testing::InitGoogleTest(); 
    return RUN_ALL_TESTS();
}

File: solutions/911/main.py
from typing import List 

class TopVotedCandidate:

    def __init__(self, persons: List[int], times: List[int]):
        self.totalVoteMap = {}
        self.latestVoteMap = {}
        self.result = []
        self.times = times
        for i in range(len(times)):
            # 计数
            if persons[i] not in self.totalVoteMap:
                self.totalVoteMap[persons[i]] = 0
            if persons[i] not in self.latestVoteMap:
                self.latestVoteMap[persons[i]] = 0
            self.totalVoteMap[persons[i]] += 1
            self.latestVoteMap[persons[i]] = times[i]
            # 唱票
            mvp, mv = 0, -float("inf")
            for p, v in self.totalVoteMap.items():
                if v > mv:
                    mvp = p
                    mv = v
                elif v == mv:
                    # 处理平局
                    if self.latestVoteMap[p] > self.latestVoteMap[mvp]:
                        mvp = p
            self.result.append(mvp)

    def q(self, t: int) -> int:
        left, right = 0, len(self.times) - 1
        while left <= right:
            mid = int(left + (right - left) / 2)
            if self.times[mid] > t:
                right = mid - 1
            elif self.times[mid] < t:
                left = mid + 1
            else:
                return self.result[mid]
        if self.times[mid] > t:
            mid -= 1
        return self.result[mid]

if __name__ == "__main__": 
    persons = [0, 1, 1, 0, 0, 1, 0]
    times = [0, 5, 10, 15, 20, 25, 30]
    obj = TopVotedCandidate(persons, times)
    queries = [3, 12, 25, 15, 24, 8]
    ans = [0, 1, 1, 0, 0, 1]
    for i in range(len(queries)): 
        assert obj.q(queries[i]) == ans[i] 
